java-checks/src/main/java/org/sonar/java/checks/spring/SpringBeansShouldBeAccessibleCheck.java

```diff
@@ -97,12 +99,28 @@ public void visitNode(Tree tree) {
     if (componentScanValues != null) {
       componentScanValues.forEach(this::addToScannedPackages);
     } else if (hasAnnotation(classSymbolMetadata, SPRING_BOOT_APP_ANNOTATION)) {
-      packagesScannedBySpring.add(classPackageName);
+      packagesScannedBySpring.addAll(targetedPackages(classPackageName, classSymbolMetadata));
     } else if (hasAnnotation(classSymbolMetadata, SPRING_BEAN_ANNOTATIONS)) {
       addMessageToMap(classPackageName, classTree.simpleName());
     }
   }
 
+  private static List<String> targetedPackages(String classPackageName, SymbolMetadata classSymbolMetadata) {
+    // annotation is necessarily there already
+    return Objects.requireNonNull(classSymbolMetadata.valuesForAnnotation(SPRING_BOOT_APP_ANNOTATION))
+      .stream()
+      .filter(v -> "scanBasePackages".equals(v.name()))
+      .map(SymbolMetadata.AnnotationValue::value)
+      .findFirst()
+      // list of packages to scan
+      .filter(Object[].class::isInstance)
+      .map(Object[].class::cast)
+      .map(Arrays::asList)
+      .map(list -> list.stream().map(String.class::cast).collect(Collectors.toList()))
```

Comment:
It seems unlikely to happen, but the values of annotations can be Identifiers, making this cast unsafe. Example:
```
class MyClass {
  private final String PACKAGE = "my.package";

  @SpringBootApplication(scanBasePackages = PACKAGE)
  public class SpringBootApp {
  }
}
```

Reply:
good catch

Reply:
In fact nop... I forgot about it but we are not relying anymore on the tree when getting annotation values... but directly on the semantic engine. So it's always going to be an `Object[]` filled with `String`s. I anyway added a test to prove it.

Reply:
As discussed, most of the time it will work, but there are still corner cases where the object could be something other than a String.